proof/refine/ARM/PageTableDuplicates.thy

```diff
@@ -1168,38 +1166,46 @@ lemma createObject_valid_duplicates'[wp]:
   apply (intro conjI impI allI)
-      apply simp
+        apply simp
+       apply clarsimp
+       apply (drule(2) valid_duplicates'_update) prefer 3
+         apply fastforce
+        apply (simp add: vs_entry_align_def)
+       apply simp
+      apply clarsimp
+      apply (drule(2) valid_duplicates'_update) prefer 3
+        apply (fastforce simp: vs_entry_align_def)+
      apply clarsimp
      apply (drule(2) valid_duplicates'_update) prefer 3
-       apply fastforce
-      apply (simp add: vs_entry_align_def)
-     apply simp
-    apply clarsimp
-    apply (drule(2) valid_duplicates'_update) prefer 3
-      apply (fastforce simp: vs_entry_align_def)+
+       apply (fastforce simp: vs_entry_align_def)+
+    apply (clarsimp simp:ARM_H.toAPIType_def word_bits_def
+                   split:ARM_H.object_type.splits)
+    apply (cut_tac ptr = ptr in new_cap_addrs_fold'[where n = "2^us"
+               and ko = "(KOCTE makeObject)",simplified])
+     apply (rule word_1_le_power)
+     apply (clarsimp simp: word_bits_def)
+    apply (drule_tac ptr = ptr and ko = "KOCTE makeObject" in
+           valid_duplicates'_insert_ko[where us = us,simplified])
+        apply (simp add: APIType_capBits_def is_aligned_mask toAPIType_def
+                  split: ARM_H.object_type.splits)
+       apply (simp add: vs_entry_align_def)
+      apply (simp add: objBits_simps')
+     apply (rule none_in_new_cap_addrs
+            ,(simp add: objBits_simps' pageBits_def APIType_capBits_def
+                        ARM_H.toAPIType_def
+                        word_bits_conv archObjSize_def is_aligned_mask
+                 split: ARM_H.object_type.splits)+)[1]
+    apply (clarsimp simp: word_bits_def)
+   apply simp
    apply clarsimp
    apply (drule(2) valid_duplicates'_update) prefer 3
-     apply (fastforce simp: vs_entry_align_def)+
-  apply (clarsimp simp:ARM_H.toAPIType_def word_bits_def
-                 split:ARM_H.object_type.splits)
-  apply (cut_tac ptr = ptr in new_cap_addrs_fold'[where n = "2^us"
-   and ko = "(KOCTE makeObject)",simplified])
-   apply (rule word_1_le_power)
-  apply (clarsimp simp: word_bits_def)
-  apply (drule_tac ptr = ptr and ko = "KOCTE makeObject" in
-    valid_duplicates'_insert_ko[where us = us,simplified])
-      apply (simp add: APIType_capBits_def is_aligned_mask toAPIType_def
-                split: ARM_H.object_type.splits)
-     apply (simp add: vs_entry_align_def)
-   apply (simp add: objBits_simps')
-   apply (rule none_in_new_cap_addrs
-     ,(simp add: objBits_simps' pageBits_def APIType_capBits_def
-                 ARM_H.toAPIType_def
-                 word_bits_conv archObjSize_def is_aligned_mask
-          split: ARM_H.object_type.splits)+)[1]
-  apply (clarsimp simp: word_bits_def)*)
- sorry
-
+     apply fastforce
+    apply (simp add: vs_entry_align_def)
+   apply simp
+  apply clarsimp
+  apply (drule(2) valid_duplicates'_update) prefer 3
```

Comment:
nitpick: This prefer 3 is not necessary?

Reply:
No bueno

Reply:
Interesting! Perhaps there is some kind of schematic unification that has to happen in a particular order. Cool, no worries.

Reply:
Actually, `apply (fastforce elim!: valid_duplicates'_update simp: vs_entry_align_def)` works fine instead.